contrib: fix check-deps.sh to check for weak symbols

ryanofsky opened the pull request:
Fix check-deps.sh to check for weak symbols so it can detect when an exported template function like is used from another library.

Also update the script to work with cmake and configure it to run as part of CI.

Problem was reported by hebasto in https://github.com/bitcoin/bitcoin/pull/29015#issuecomment-2209258843

ryanofsky committed:
contrib: fix check-deps.sh to check for weak symbols

Fix check-deps.sh to check for weak symbols so it can detect when an exported
template function is used from another library.

In a previous version of this commit, this change caused an invalid dependency
in the consensus library on the TryParseHex template function from the util
library to be detected, and a suppression was added here. But #30377 removed
the invalid dependency so the suppression is no longer needed.

The invalid dependency and problem detecting weak symbol usage was originally
reported by Hennadii Stepanov in
https://github.com/bitcoin/bitcoin/pull/29015#issuecomment-2209258843

hebasto commented:
Concept ACK.

hebasto commented on contrib/devtools/check-deps.sh:
```diff
@@ -58,6 +58,10 @@ SUPPRESS["libbitcoin_common_a-common.o libbitcoin_node_a-interface_ui.o _Z9InitE
 # common library. It should be moved to the node library instead.
 SUPPRESS["libbitcoin_common_a-external_signer.o libbitcoin_node_a-server.o _ZN9CRPCTable13appendCommandERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEPK11CRPCCommand"]=1
+# pubkey.cpp in consensus library currently calls ParseHex in util to set
+# NUMS_H_DATA constant, but this dependency can be eliminated in #30377 so
```
nit: Maybe make a reference to a PR consistent with the [previous one](https://github.com/bitcoin/bitcoin/blob/9adebe145557ef410964dd48a02f3d239f488cd0/contrib/devtools/check-deps.sh#L55)?

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30415#discussion_r1672501712

> nit: Maybe make a reference to a PR consistent with the [previous one](https://github.com/bitcoin/bitcoin/blob/9adebe145557ef410964dd48a02f3d239f488cd0/contrib/devtools/check-deps.sh#L55)?

Thanks, updated

hebasto reviewed (approved):
ACK 5a8b9432cde11f6140855717af195d8b7e798d75, tested on Ubuntu 24.04/

ryanofsky reviewed (commented):
Updated 5a8b9432cde11f6140855717af195d8b7e798d75 -> 114b2a406e604747bd856f566aa8c7ad84dd8f15 ([`pr/weakcheck.1`](https://github.com/ryanofsky/bitcoin/commits/pr/weakcheck.1) -> [`pr/weakcheck.2`](https://github.com/ryanofsky/bitcoin/commits/pr/weakcheck.2), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/weakcheck.1..pr/weakcheck.2)) making comment more consistent

hebasto reviewed (approved):
re-ACK 114b2a406e604747bd856f566aa8c7ad84dd8f15.

ryanofsky commented:
If anyone else would like to review or test this, you can confirm that it works by just running `./contrib/devtools/check-deps.sh` in the source directory. The script will automatically call `make` to build the bitcoin libraries and should show `Success! No unexpected dependencies were detected.` if successful.

hebasto commented:
cc @m3dwards :)

fanquake commented:
Seems like if a script such as this is going to exist, it should be getting run as part of the CI, so that regressions are actually caught (if that's what we want to do)?

note: `libbitcoin_crypto_arm_shani` is missing from `LIBS[crypto]` crypto.

ryanofsky committed:
ci: run check-deps.sh as part of clang-tidy job

ryanofsky commented:
> Seems like if a script such as this is going to exist, it should be getting run as part of the CI, so that regressions are actually caught (if that's what we want to do)?

I think it would be nice to run as part of CI. I added a commit to try that.

If we think the script is a maintenance burden, it is also not essential to keep. I wrote it mainly to help write #29015 to figure out which parts of the util library should be moved to common, crypto, and consensus libraries.

> note: `libbitcoin_crypto_arm_shani` is missing from `LIBS[crypto]` crypto.

Not sure if this might be expected but when I try that on my system I see:

```
crypto/sha256_arm_shani.cpp:16:10: fatal error: arm_acle.h: No such file or directory
   16 | #include <arm_acle.h>
```

---

Added 1 commits 114b2a406e604747bd856f566aa8c7ad84dd8f15 -> dde57405e99c00dcf511ebedd42deb2de5e67e5f ([`pr/weakcheck.2`](https://github.com/ryanofsky/bitcoin/commits/pr/weakcheck.2) -> [`pr/weakcheck.3`](https://github.com/ryanofsky/bitcoin/commits/pr/weakcheck.3), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/weakcheck.2...pr/weakcheck.3)) to try to run check-deps script in CI

hebasto commented:
> If we think the script is a maintenance burden, it is also not essential to keep. I wrote it mainly to help write #29015 to figure out which parts of the util library should be moved to common, crypto, and consensus libraries.

I think it is worth having this script to prevent re-introducing undocumented inter-library dependencies.

ryanofsky commented:
Updated dde57405e99c00dcf511ebedd42deb2de5e67e5f -> 07509a4b691b4250c8e085d247dad9ddb4e5baa5 ([`pr/weakcheck.3`](https://github.com/ryanofsky/bitcoin/commits/pr/weakcheck.3) -> [`pr/weakcheck.4`](https://github.com/ryanofsky/bitcoin/commits/pr/weakcheck.4), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/weakcheck.3..pr/weakcheck.4)) to fix ci failure "/ci_container_base/ci/test/03_test_script.sh: line 155: contrib/devtools/check-deps.sh: No such file or directory" [predicted](https://github.com/bitcoin/bitcoin/pull/30415#discussion_r1691519294) by maflcko

ryanofsky committed:
contrib: fix check-deps.sh when libraries do not import symbols

Script was failing when called on libraries that do not import symbols, because
bash pipefail option was specified, and grep was used in some pipelines to
filter symbols, and grep returns status 1 when it doesn't match any lines. This
could cause the script to fail on some systems and configurations, such as the
clang-tidy CI configuration
https://cirrus-ci.com/task/4801670352207872?logs=ci#L6191 where the
libbitcoin_crypto_x86_shani.a library does not import symbols.

hebasto reviewed (approved):
re-ACK 2bca4af323c5b505689c5203e6ddba8becac5dc8

m3dwards commented:
ACK 2bca4af323c5b505689c5203e6ddba8becac5dc8

Tested on Debian x86.

With the suppression disabled I got:

```shell
Error: libbitcoin_consensus_a-pubkey.o depends on libbitcoin_util_a-strencodings.o symbol 'std::optional<std::vector<unsigned char, std::allocator<unsigned char> > > TryParseHex<unsigned char>(std::basic_string_view<char, std::char_traits<char> >)'
```

hebasto commented:
cc @maflcko @fanquake

ryanofsky committed:
contrib: make check-deps.sh script work with cmake

ryanofsky commented:
Rebased 2bca4af323c5b505689c5203e6ddba8becac5dc8 -> 76ef8113f70070ab1deeeb142977d46d8132c36e ([`pr/weakcheck.7`](https://github.com/ryanofsky/bitcoin/commits/pr/weakcheck.7) -> [`pr/weakcheck.8`](https://github.com/ryanofsky/bitcoin/commits/pr/weakcheck.8), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/weakcheck.7-rebase..pr/weakcheck.8)) updating this to work with cmake and dropping suppression no longer needed after #30377

hodlinator commented on contrib/devtools/check-deps.sh:
```diff
@@ -32,32 +31,28 @@ ALLOWED_DEPENDENCIES=(
     "wallet common"
     "wallet crypto"
     "wallet util"
-    "wallet_tool util"
-    "wallet_tool wallet"
 )
 
 # Add minor dependencies omitted from doc/design/libraries.md to keep the
 # dependency diagram simple.
 ALLOWED_DEPENDENCIES+=(
     "wallet consensus"
-    "wallet_tool common"
-    "wallet_tool crypto"
```
What happened to **wallet_tool** in the latest push?

2 dependencies exist according to **/doc/design/libraries.md**: 
```
libbitcoin_wallet_tool-->libbitcoin_wallet;
libbitcoin_wallet_tool-->libbitcoin_util;
```

ryanofsky replied:
I didn't dig into it but it appeared wallet_tool library disappeared in the conversion to cmake, so I removed references to it here. Probably libraries.md needs to be updated too though. There may be an existing cmake documentation PR where that change could be / has been made

hebasto replied:
> I didn't dig into it but it appeared wallet_tool library disappeared in the conversion to cmake...

I couldn’t bring myself to create a library for just a single object file. :)

For the reference, the library was defined as follows: https://github.com/bitcoin/bitcoin/blob/80f00cafdeef0600fa1a5e906686720786c2336c/src/Makefile.am#L547-L549

ryanofsky commented:
Updated 76ef8113f70070ab1deeeb142977d46d8132c36e -> 3e4312eef78f233eb7ae1d7d85e497de34144f2e ([`pr/weakcheck.8`](https://github.com/ryanofsky/bitcoin/commits/pr/weakcheck.8) -> [`pr/weakcheck.9`](https://github.com/ryanofsky/bitcoin/commits/pr/weakcheck.9), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/weakcheck.8..pr/weakcheck.9)) to fix failing CI job due to changed default build directory (https://cirrus-ci.com/task/5141392714891264?logs=ci#L5325)

TheCharlatan commented on contrib/devtools/check-deps.sh:
```diff
@@ -194,6 +191,7 @@ cd "$BUILD_DIR"
 # shellcheck disable=SC2046
 make -j"$(nproc)" $(lib_targets)
```
I think we are trying to consistently use `cmake --build` over `make `.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30415#discussion_r1745115109

> I think we are trying to consistently use `cmake --build` over `make `.

Thanks, switched to cmake --build

sedited reviewed (approved):
ACK 3e4312eef78f233eb7ae1d7d85e497de34144f2e

ryanofsky reviewed (commented):
Updated 3e4312eef78f233eb7ae1d7d85e497de34144f2e -> b14d87085fca777b1d14ff051d31d41932597f06 ([`pr/weakcheck.9`](https://github.com/ryanofsky/bitcoin/commits/pr/weakcheck.9) -> [`pr/weakcheck.10`](https://github.com/ryanofsky/bitcoin/commits/pr/weakcheck.10), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/weakcheck.9..pr/weakcheck.10)) with suggested change to use cmake --build wrapper instead of calling make directly

sedited reviewed (approved):
Re-ACK b14d87085fca777b1d14ff051d31d41932597f06

ryanofsky commented:
Updated b14d87085fca777b1d14ff051d31d41932597f06 -> 3ae35b427fe59bc9ab24d07c1adb46faa702de20 ([`pr/weakcheck.10`](https://github.com/ryanofsky/bitcoin/commits/pr/weakcheck.10) -> [`pr/weakcheck.11`](https://github.com/ryanofsky/bitcoin/commits/pr/weakcheck.11), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/weakcheck.10..pr/weakcheck.11)) simplifying by dropping `cd` since cmake --build ignores the current directory.

sedited reviewed (approved):
Re-ACK 3ae35b427fe59bc9ab24d07c1adb46faa702de20

hebasto reviewed (approved):
ACK 3ae35b427fe59bc9ab24d07c1adb46faa702de20, I have reviewed the code and it looks OK. Also I've tested it locally.

fanquake commented:
Fixing this looks good. I guess going forward it currently has the caveat that it's expected to work most correctly in the CI (on the same hardware used by the CI). i.e this it fail for any dev trying to run the script locally, on aarch64:
```bash
[ 25%] Built target bitcoin_crypto_arm_shani
[100%] Built target bitcoin_crypto
gmake: *** No rule to make target 'bitcoin_crypto_x86_shani'.  Stop.
```
or if a dev compiles on x86_64, but doesn't match the CI config, they may see incorrect output i.e:
```bash
[100%] Built target bitcoin_crypto_sse41
Built target bitcoin_crypto_avx2
Warning: suppression 'init.cpp.o bdb.cpp.o _ZN6wallet27BerkeleyDatabaseSanityCheckEv' was ignored, consider deleting.
Success! No unexpected dependencies were detected.
```

This seems fine for now, as long as weak symbols/regressions are caught, and could be followed up on.

hebasto commented on contrib/devtools/check-deps.sh:
```diff
@@ -67,8 +62,10 @@ usage() {
+          local name="${lib_path##*/}"
+          name="${name#lib}"
+          name="${name%.a}"
+          echo "$name"
```
The comment above no longer seems correct.

hebasto replied:
Adjusted in https://github.com/bitcoin/bitcoin/pull/30901.

fanquake commented:
Looks like this doesn't fail when it detects a problem? i.e the tidy job is passing (most recent run: https://api.cirrus-ci.com/v1/task/5757501226876928/logs/ci.log), but is currently outputting:
```bash
[21:19:24.426] Built target bitcoin_crypto_avx2
[21:19:26.500] Error: netif.cpp.o depends on netaddress.cpp.o symbol 'CNetAddr::CNetAddr(in_addr const&)', can suppess with:
[21:19:26.500]     SUPPRESS["netif.cpp.o netaddress.cpp.o _ZN8CNetAddrC1ERK7in_addr"]=1
[21:19:26.502] Error: pcp.cpp.o depends on netaddress.cpp.o symbol 'CNetAddr::CNetAddr(in_addr const&)', can suppess with:
[21:19:26.502]     SUPPRESS["pcp.cpp.o netaddress.cpp.o _ZN8CNetAddrC1ERK7in_addr"]=1
[21:19:26.516] Error: netif.cpp.o depends on netaddress.cpp.o symbol 'CNetAddr::CNetAddr(in6_addr const&, unsigned int)', can suppess with:
[21:19:26.516]     SUPPRESS["netif.cpp.o netaddress.cpp.o _ZN8CNetAddrC1ERK8in6_addrj"]=1
[21:19:26.517] Error: pcp.cpp.o depends on netaddress.cpp.o symbol 'CNetAddr::CNetAddr(in6_addr const&, unsigned int)', can suppess with:
[21:19:26.517]     SUPPRESS["pcp.cpp.o netaddress.cpp.o _ZN8CNetAddrC1ERK8in6_addrj"]=1
[21:19:26.530] Error: pcp.cpp.o depends on netaddress.cpp.o symbol 'CService::SetSockAddr(sockaddr const*)', can suppess with:
[21:19:26.530]     SUPPRESS["pcp.cpp.o netaddress.cpp.o _ZN8CService11SetSockAddrEPK8sockaddr"]=1
[21:19:26.542] Error: pcp.cpp.o depends on netaddress.cpp.o symbol 'CService::CService(in_addr const&, unsigned short)', can suppess with:
[21:19:26.542]     SUPPRESS["pcp.cpp.o netaddress.cpp.o _ZN8CServiceC1ERK7in_addrt"]=1
[21:19:26.554] Error: pcp.cpp.o depends on netaddress.cpp.o symbol 'CService::CService(CNetAddr const&, unsigned short)', can suppess with:
[21:19:26.554]     SUPPRESS["pcp.cpp.o netaddress.cpp.o _ZN8CServiceC1ERK8CNetAddrt"]=1
[21:19:26.567] Error: pcp.cpp.o depends on netaddress.cpp.o symbol 'CService::CService()', can suppess with:
[21:19:26.567]     SUPPRESS["pcp.cpp.o netaddress.cpp.o _ZN8CServiceC1Ev"]=1
[21:19:26.579] Error: pcp.cpp.o depends on netaddress.cpp.o symbol 'CNetAddr::GetIn6Addr(in6_addr*) const', can suppess with:
[21:19:26.579]     SUPPRESS["pcp.cpp.o netaddress.cpp.o _ZNK8CNetAddr10GetIn6AddrEP8in6_addr"]=1
[21:19:26.590] Error: pcp.cpp.o depends on netaddress.cpp.o symbol 'CNetAddr::ToStringAddr[abi:cxx11]() const', can suppess with:
[21:19:26.590]     SUPPRESS["pcp.cpp.o netaddress.cpp.o _ZNK8CNetAddr12ToStringAddrB5cxx11Ev"]=1
[21:19:26.603] Error: pcp.cpp.o depends on netaddress.cpp.o symbol 'CNetAddr::GetInAddr(in_addr*) const', can suppess with:
[21:19:26.603]     SUPPRESS["pcp.cpp.o netaddress.cpp.o _ZNK8CNetAddr9GetInAddrEP7in_addr"]=1
[21:19:26.617] Error: netif.cpp.o depends on netaddress.cpp.o symbol 'CNetAddr::IsBindAny() const', can suppess with:
[21:19:26.617]     SUPPRESS["netif.cpp.o netaddress.cpp.o _ZNK8CNetAddr9IsBindAnyEv"]=1
[21:19:26.631] Error: pcp.cpp.o depends on netaddress.cpp.o symbol 'CService::GetSockAddr(sockaddr*, unsigned int*) const', can suppess with:
[21:19:26.631]     SUPPRESS["pcp.cpp.o netaddress.cpp.o _ZNK8CService11GetSockAddrEP8sockaddrPj"]=1
[21:19:26.645] Error: pcp.cpp.o depends on netaddress.cpp.o symbol 'CService::ToStringAddrPort[abi:cxx11]() const', can suppess with:
[21:19:26.645]     SUPPRESS["pcp.cpp.o netaddress.cpp.o _ZNK8CService16ToStringAddrPortB5cxx11Ev"]=1
[21:19:26.669] Error: Unexpected dependencies were detected. Check previous output.
```

ryanofsky commented:
> Looks like this doesn't fail when it detects a problem? i.e the tidy job is passing (most recent run: https://api.cirrus-ci.com/v1/task/5757501226876928/logs/ci.log), but is currently outputting:

Nice catch. Opened #31011 to address this
